### DISCUSSION TITLE
Converting Number integer to float to calculating with them, than converting it back and Output doesn't work

### DISCUSSION THREAD

**Discussion by Guelakais** [Rahix/avr-hal#307] [Q&A] [answered]
Sorry to ask you again. I wanted to work with a temperature sensor. First I've get outputs over thousand so i thought, my output minus thousand would be the right temperature. Now one of my colleges hast told me, this is wrong and I need more calculations.
`
#![no_std]
#![no_main]

use panic_halt as _;
use arduino_hal::prelude::*;

#[arduino_hal::entry]
fn main() -> ! {
    let dp = arduino_hal::Peripherals::take().unwrap();
    let pins = arduino_hal::pins!(dp);

    let mut serial = arduino_hal::default_serial!(dp, pins, 57600);
    let mut adc = arduino_hal::Adc::new(dp.ADC, Default::default());

    let a0 = pins.a0.into_analog_input(&mut adc);

    loop {
        let val = ((a0.analog_read(&mut adc) as f32)/37.9) as i8;
        ufmt::uwrite!(&mut serial, "{};", val).void_unwrap();
    }
}
`
Now, I get very cryptic error messages. Everything works fine, until I uncomment 
`
ufmt::uwrite!(&mut serial, "{};", val).void_unwrap();
`
Do you've ever had a similar issue? Do you know, how to solve it?

**Comment by stappersg** [comment]
Please, pretty please, properly format your question.  (Visit this #307 and edit it)

Karma bonus points for showing those _very cryptic error messages_.

**Reply by Rahix** [reply]
As @stappersg said, you need to provide the error message.  Without it, we have no chance of understanding what the problem is.  Just because it looks cryptic to you does not mean it doesn't contain valuable details about the problem :)

**Reply by Guelakais** [reply]
Ah, you're right. Sorry. Here the Message:
```
cargo build     
   Compiling arduino-uno v0.1.0 (/home/fiolakais/projects/RustArduino/arduino-uno)
error: linking with `avr-gcc` failed: exit status: 1
  |
  = note: "avr-gcc" "-mmcu=atmega328p" "-Wl,--as-needed" "/home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/temp_sensor-96e0a98d8f484dd9.arduino_hal-eca00398bbed5ea2.arduino_hal.da33cd1b-cgu.0.rcgu.o.rcgu.o" "-Wl,--as-needed" "-L" "/home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps" "-L" "/home/fiolakais/projects/RustArduino/arduino-uno/target/debug/deps" "-L" "/home/fiolakais/.rustup/toolchains/nightly-2022-07-10-x86_64-unknown-linux-gnu/lib/rustlib/avr-atmega328p/lib" "-Wl,--start-group" "-Wl,--end-group" "-Wl,-Bstatic" "/home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/libcompiler_builtins-b0ea9bb28a6c4367.rlib" "-Wl,-Bdynamic" "-lgcc" "-Wl,-znoexecstack" "-L" "/home/fiolakais/.rustup/toolchains/nightly-2022-07-10-x86_64-unknown-linux-gnu/lib/rustlib/avr-atmega328p/lib" "-o" "/home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/temp_sensor-96e0a98d8f484dd9.elf" "-Wl,--gc-sections" "-no-pie"
  = note: /usr/bin/avr-ld: /home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/libcompiler_builtins-b0ea9bb28a6c4367.rlib(compiler_builtins-b0ea9bb28a6c4367.compiler_builtins.9cb26a38-cgu.15.rcgu.o): in function `<f32 as compiler_builtins::float::Float>::normalize':
          /home/fiolakais/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.73/src/int/specialized_div_rem/binary_long.rs:(.text._ZN55_$LT$f32$u20$as$u20$compiler_builtins..float..Float$GT$9normalize17hda9c0c14a7372049E+0x182): undefined reference to `core::panicking::panic'
          /usr/bin/avr-ld: /home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/libcompiler_builtins-b0ea9bb28a6c4367.rlib(compiler_builtins-b0ea9bb28a6c4367.compiler_builtins.9cb26a38-cgu.2.rcgu.o): in function `compiler_builtins::float::div::div32':
          /home/fiolakais/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.73/src/float/div.rs:(.text._ZN17compiler_builtins5float3div8__divsf317ha78b8af2ee058d4fE+0x724): undefined reference to `core::panicking::panic'
          /usr/bin/avr-ld: /home/fiolakais/projects/RustArduino/arduino-uno/target/avr-atmega328p/debug/deps/libcompiler_builtins-b0ea9bb28a6c4367.rlib(compiler_builtins-b0ea9bb28a6c4367.compiler_builtins.9cb26a38-cgu.8.rcgu.o): in function `compiler_builtins::float::conv::int_to_float::u32_to_f32_bits':
          /home/fiolakais/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.73/src/float/conv.rs:15: undefined reference to `core::panicking::panic'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `arduino-uno` due to previous error
```

**Comment by stappersg** [comment]
Heart of the _cryptic error message_ is
```text
$ cargo build
error: linking with `avr-gcc` failed: exit status: 1
$
```
@Guelakais in #305 you had a succesfull compile.  Have a look over there and tell us what you are doing different for this issue.   That request in other words.  I think you have already a working tool chain. That it is a good thing to compare the _known good_, `#305`  and the _not yet working_, `#307`.

**Comment by Rahix** [comment, marked as answer]
@Guelakais, right, you ran into a known problem there.  See https://github.com/rust-lang/compiler-builtins/issues/347#issuecomment-706403824 for the upstream report and a workaround.

**Reply by Guelakais** [reply]
Thanks.
```
[profile.dev.package.compiler_builtins]
overflow-checks = false
```
Did solve the problem for the compiler. However I'll only know as the code really works, if I'm tomorrow back into the laboratory to run it on the Arduino.
